src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs

```diff
@@ -91,10 +92,27 @@ public override void OnNodeModified(NodeModel obj)
 
         private void hwm_SetNodeDeltaState(object sender, DeltaComputeStateEventArgs e)
         {
-            var nodeGuids = e.NodeGuidList;
+            // Using Nodes here is not thread safe, as nodes can be added/removed by the UI thread midway.
+            // Dispatching this to the UI thread helps avoid concurrency issues and seems more correct.
+            if (!DynamoViewModel.Model.ShutdownRequested)
+            {
+                DynamoViewModel.UIDispatcher.Invoke(() => UpdateNodesDeltaState(e.NodeGuidList, e.GraphExecuted));
+            }
+            else
+            {
+                // However, when Dynamo is shutting down, a funny deadlock situation occurs.
+                // The main thread is waiting for the scheduler to finish executing.
+                // So in order to avoid it, we do the update on the scheduler thread.
+                // Concurrency should not be an issue now that the UI is blocked waiting for shutdown.
+                UpdateNodesDeltaState(e.NodeGuidList, e.GraphExecuted);
```

Comment:
I'm curious about this because I thought `Nodes` had a lock around it? This is not sufficient?

Reply:
I didn't see a lock around here. Maybe you are thinking about the Model?

Reply:
yes, thats what I was thinking of. ✔️